radix-engine-tests/tests/wasm_buffers.rs

```diff
@@ -0,0 +1,486 @@
+        (10 * KB, 10 * KB + 127 * KB, 1 * KB),
+        // Large buffers
+        (1 * MB, 0, 1 * MB + 128 * KB),
+    ] {
+        // Act
+        let receipt = test_wasm_buffer_read!(
+            test_runner,
+            component_address,
+            read = (buffer_size, memory_offs, memory_len)
+        );
+
+        // Assert
+        receipt.expect_specific_failure(|e| {
+            matches!(
+                e,
+                RuntimeError::VmError(VmError::Wasm(WasmRuntimeError::MemoryAccessError)),
+            )
+        });
+    }
+}
+
+#[test]
+fn test_wasm_buffers_write_memory_access_error() {
+    // Arrange
+    let (mut test_runner, component_address) = get_test_runner();
+
+    for ((buffer_size, memory_offs, memory_len), (write_buffer_size, write_memory_offs)) in [
+        // Add 128KB to write memory offs to make sure we are accessing beyond WASM memory
+        (
+            (10 * KB, 0, get_memory_len(10 * KB)),
+            (10 * KB, 10 * KB + 128 * KB),
+        ),
+        (
+            (1 * MB, 0, get_memory_len(1 * MB)),
+            (1 * MB, 1 * MB + 128 * KB),
+        ),
+    ] {
+        // Act
+        let receipt = test_wasm_buffer_read_write!(
+            test_runner,
+            component_address,
+            read = (buffer_size, memory_offs, memory_len),
+            write = (write_buffer_size, write_memory_offs)
+        );
+
+        // Assert
+        receipt.expect_specific_failure(|e| {
+            matches!(
+                e,
+                RuntimeError::VmError(VmError::Wasm(WasmRuntimeError::MemoryAccessError)),
+            )
+        });
+    }
+}
+
+#[test]
+fn test_wasm_buffer_read_memory_substate_size_exceeded() {
+    // Arrange
+    let (mut test_runner, component_address) = get_test_runner();
+
+    // Act
+    let receipt = test_wasm_buffer_read!(
+        test_runner,
+        component_address,
+        read = (
+            MAX_SUBSTATE_VALUE_SIZE as u64,
+            0,
+            get_memory_len(MAX_SUBSTATE_VALUE_SIZE as u64)
+        )
+    );
+
+    // Assert
+    receipt.expect_specific_failure(|e| {
+        matches!(
+            e,
+            RuntimeError::SystemModuleError(SystemModuleError::TransactionLimitsError(
+                TransactionLimitsError::MaxSubstateSizeExceeded(..)
+            ))
+        )
+    });
+}
+
+#[test]
+fn test_wasm_buffer_read_memory_instruction_trap() {
+    // Arrange
+    let (mut test_runner, component_address) = get_test_runner();
+
+    // Act
+    let receipt = test_wasm_buffer_read!(
+        test_runner,
+        component_address,
+        read = (4 * MB, 0, get_memory_len(4 * MB))
+    );
+
+    // Assert
+    receipt.expect_specific_failure(|e| match e {
+        // This error is really nasty and we should somehow prevent it from occuring.
+        // Especially that we know that transaction will fail for smaller sizes...
+        RuntimeError::VmError(VmError::Wasm(WasmRuntimeError::ExecutionError(message))) => {
```

Comment:
I really don't like this error and I think it would good to prevent it somehow from occuring.